lib/decorateWithMetrics.js

```diff
@@ -0,0 +1,132 @@
+import _ from 'underscore';
+
+/**
+ * Each key is a method name and the value is an array of calls metadata
+ * @type {Record<String, Array<CallMeta>>}
+ */
+let stats = {};
+
+/**
+ * Class instances are easily tracked by Dev tools memory snapshots
+ * Wrapping a result with such instance allows us to track it without
+ * incurring extra weight
+ */
+class TaggedItem {
+    constructor(data) {
+        Object.assign(this, data);
+        this.ref = data;
+    }
+}
+
+/**
+ * Capture tags in a weak map
+ * When the original object (key) is no longer used anywhere in app - the data (TaggedItem)
+ * is free for garbage collection and removed
+ * Using dev tools we see which TaggedItems are still retained and what's using them - preventing GC
+ * @type {WeakMap<object, TaggedItem>}
+ */
+const taggedItems = new WeakMap();
```

Comment:
Just to get clear on this, we are using `WeakMap` because it will _allow garbage collection_. And therefore reveal to us which `TaggedItem` no longer have references. Why does this matter?

Reply:
It's actually the opposite - it would reveal to us which `TaggedItems` are still retained and allow us to see what is retaining them

A `TaggedItem` would only be retained if the `ref` - the original data is still used somewhere

We can find retained `TaggedItems` in a heap snapshot and then on the **Retainers** section we'll can track what is retaining a specific `TaggedItem` all the way to the component that is using the original (tagged) data

Let's say we see a `TaggedItem` that is retaining a huge chunk of memory - we can track what's causing this

Reply:
>it would reveal to us which TaggedItems are still retained and allow us to see what is retaining them

Sorry, can you add all of this context in the code as a comment or possibly a guide somewhere? I'm not sure how or why any of this works, why we want it, etc and if we agree it's useful then it should be properly documented. 

>Let's say we see a TaggedItem that is retaining a huge chunk of memory - we can track what's causing this

That's interesting. But this isn't the problem that we are trying to solve.

Reply:
> > it would reveal to us which TaggedItems are still retained and allow us to see what is retaining them
> 
> Sorry, can you add all of this context in the code as a comment or possibly a guide somewhere? I'm not sure how or why any of this works, why we want it, etc and if we agree it's useful then it should be properly documented.

I've made a demo and explained how that works in the top level comment when I pushed the changes: https://github.com/Expensify/react-native-onyx/pull/70#pullrequestreview-660153513



> > Let's say we see a TaggedItem that is retaining a huge chunk of memory - we can track what's causing this
> 
> That's interesting. But this isn't the problem that we are trying to solve.

Ok then there's no point documenting and spending more time over this, I'll just remove memory tracking

Reply:
Ok, I might be confused, but to clarify, is this something we agreed to do or not? If it is, then I'd suggest including documentation so people will know how to use it. If we don't need/want it, then yes, I'd remove it. You can always ask in the Slack channel for feedback.